### PR TITLE
Carnix: 0.9 -> 0.10

### PR DESCRIPTION
**Pull request by P-E-Meunier**
###### Motivation for this change

Carnix 0.10 can handle even trickier crates such as Sequoia-PGP.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/build-support/rust/crates-io.nix [hunk 1]
```diff
@@ -4,6 +4,30 @@ let inherit (lib.lists) fold;
...
+  features_.aho_corasick."0.6.10" = deps: f: updateFeatures f (rec {
+    aho_corasick."0.6.10".default = (f.aho_corasick."0.6.10".default or true);
+    memchr."${deps.aho_corasick."0.6.10".memchr}".default = true;
+  }) [
+    (features_.memchr."${deps."aho_corasick"."0.6.10"."memchr"}" deps)
+  ];
+
+
+# end
 # aho-corasick-0.6.8
```
Comment (Infinisil) [inline pkgs/build-support/rust/crates-io.nix]: Does this need multiple versions of the same packages?

**Comment by ln-nl** [comment]
Just curious: is `crates-io.nix` going to be "append only"?

**Comment by P-E-Meunier** [comment]
@ln-nl I don't know. At the moment it is, but if this becomes problematic (i.e. too big) we might want to find ways to refactor it / trim it at some point,

File: pkgs/build-support/rust/crates-io.nix (hunk 1 above)
Comment (P-E-Meunier) [reply pkgs/build-support/rust/crates-io.nix]: I believe it's pretty common for that particular dependency (aho-corasick).

Comment (P-E-Meunier) [reply pkgs/build-support/rust/crates-io.nix]: No wait, I replied too fast. No, it doesn't, but other packages might need the other version (crates-io.nix is common to all Rust packages in nixpkgs).

**Comment by dywedir** [comment]
cc @Mic92

@GrahamcOfBorg build carnix cargo-vendor

**Comment by Mic92** [comment]
cc @andir

**Review by andir — APPROVED** [review]
While I also see the issue of ever growing `crates-io.nix` we haven't come up with a better strategy just yet...

I build one of my side-projects (https://broken.sh) using a nix expression generated by the newer carnix version from this PR. Works just fine.

A while ago I thought about handling the issue with the always growing `crates-io.nix` file. To solve it and still provide all the Dependencies for all the `buildRustCrate` consumers in nixpkgs we can either

- list all of the crates on crates.io -> I think this is a bit over the top
- (re)generate the list regulary (on every change to the file?) by querying all the consumers about their dependencies and then merging them again..

For the 2nd option I hacked a PoC (https://gist.github.com/andir/c637c78450e6ae7815d98e7fa08d45dd) that illustrates my idea.. But that is something for another PR/issue/... .

**Pull request merged**